File: platform/windows/MihomoMeter.Windows.Core/Application/SharedCoreTrafficRoute.cs
namespace MihomoMeter.Windows.Core.Application;

public readonly record struct SharedCoreTrafficRouteObservation(
    SharedCoreTrafficFormat Format,
    SharedCoreTrafficRouteSource Source,
    SharedCoreTrafficRouteStatus Status);

internal sealed class SharedCoreTrafficRouteObservationGate
{
    private readonly HashSet<SharedCoreTrafficRouteObservation> _reportedObservations = [];

    public bool ShouldReport(SharedCoreTrafficRouteObservation observation)
    {
        return _reportedObservations.Add(observation);
    }

    public void Reset()
    {
        _reportedObservations.Clear();
    }
}

public static class SharedCoreTrafficRoute
{
    private static readonly object StateLock = new();
    private static readonly SharedCoreTrafficRouteObservationGate ObservationGate = new();
    private static Action<SharedCoreTrafficRouteObservation>? _reporter;

    public static void ConfigureReporter(
        Action<SharedCoreTrafficRouteObservation>? reporter)
    {
        lock (StateLock)
        {
            _reporter = reporter;
            ObservationGate.Reset();
        }
    }

    public static string Resolve(
        ulong bytes,
        string nativeText,
        SharedCoreTrafficFormat format)
    {
        return Resolve(
            bytes,
            nativeText,
            format,
            () => MihomoMeterSharedCore.AbiVersion,
            MihomoMeterSharedCore.ScaleTraffic);
    }

    internal static string Resolve(
        ulong bytes,
        string nativeText,
        SharedCoreTrafficFormat format,
        Func<uint> abiVersion,
        Func<ulong, SharedTrafficScale> scaleTraffic)
    {
        var result = SharedCoreTrafficRouter.Route(
            bytes,
            nativeText,
            format,
            abiVersion,
            scaleTraffic);
        var observation = new SharedCoreTrafficRouteObservation(
            format,
            result.Source,
            RouteStatus(result.Status));
        Report(observation);
        return result.Text;
    }

    public static string ResolveLazy(
        ulong bytes,
        Func<string> nativeFallback,
        SharedCoreTrafficFormat format)
    {
        return ResolveLazy(
            bytes,
            nativeFallback,
            format,
            () => MihomoMeterSharedCore.AbiVersion,
            MihomoMeterSharedCore.ScaleTraffic);
    }

    internal static string ResolveLazy(
        ulong bytes,
        Func<string> nativeFallback,
        SharedCoreTrafficFormat format,
        Func<uint> abiVersion,
        Func<ulong, SharedTrafficScale> scaleTraffic)
    {
        var result = SharedCoreTrafficRouter.RouteLazy(
            bytes,
            nativeFallback,
            format,
            abiVersion,
            scaleTraffic);
        Report(
            new SharedCoreTrafficRouteObservation(
                format,
                result.Source,
                result.Status));
        return result.Text;
    }

    private static void Report(SharedCoreTrafficRouteObservation observation)
    {
        Action<SharedCoreTrafficRouteObservation>? reporter;
        lock (StateLock)
        {
            reporter = _reporter is not null && ObservationGate.ShouldReport(observation)
                ? _reporter
                : null;
        }
        try
        {
            reporter?.Invoke(observation);
        }
        catch
        {
            // 路由诊断不得改变共享主路径或原生回退结果。
        }
    }

    private static SharedCoreTrafficRouteStatus RouteStatus(
        SharedCoreTrafficShadowStatus shadowStatus)
    {
        return shadowStatus switch
        {
            SharedCoreTrafficShadowStatus.Matched => SharedCoreTrafficRouteStatus.Matched,
            SharedCoreTrafficShadowStatus.AbiMismatch =>
                SharedCoreTrafficRouteStatus.AbiMismatch,
            SharedCoreTrafficShadowStatus.NativeCallFailed =>
                SharedCoreTrafficRouteStatus.NativeCallFailed,
            SharedCoreTrafficShadowStatus.UnexpectedResult =>
                SharedCoreTrafficRouteStatus.UnexpectedResult,
            SharedCoreTrafficShadowStatus.Mismatch => SharedCoreTrafficRouteStatus.Mismatch,
            SharedCoreTrafficShadowStatus.UnknownFailure =>
                SharedCoreTrafficRouteStatus.UnknownFailure,
            _ => SharedCoreTrafficRouteStatus.UnknownFailure,
        };
    }
}

File: Sources/Application/SharedCoreTrafficRoute.swift
import os

struct SharedCoreTrafficRouteObservation: Hashable, Sendable {
  let format: SharedCoreTrafficFormat
  let source: SharedCoreTrafficRouteSource
  let status: SharedCoreTrafficRouteStatus
}

private struct SharedCoreTrafficRouteObservationGate: Sendable {
  private var reportedObservations: Set<SharedCoreTrafficRouteObservation> = []

  mutating func shouldReport(_ observation: SharedCoreTrafficRouteObservation) -> Bool {
    reportedObservations.insert(observation).inserted
  }

  mutating func reset() {
    reportedObservations.removeAll(keepingCapacity: true)
  }
}

enum SharedCoreTrafficRoute {
  typealias Reporter = @Sendable (SharedCoreTrafficRouteObservation) throws -> Void

  private struct State {
    var reporter: Reporter?
    var observationGate = SharedCoreTrafficRouteObservationGate()
  }

  private static let stateLock = OSAllocatedUnfairLock(initialState: State())

  static func configure(reporter: Reporter?) {
    stateLock.withLock { state in
      state.reporter = reporter
      state.observationGate.reset()
    }
  }

  static func resolve(
    bytes: UInt64,
    nativeText: String,
    format: SharedCoreTrafficFormat,
    scaleTraffic: (UInt64) throws -> SharedTrafficScale =
      MihomoMeterSharedCoreAdapter.scaleTraffic(bytes:)
  ) -> String {
    let result = SharedCoreTrafficRouter.route(
      bytes: bytes,
      nativeText: nativeText,
      format: format,
      scaleTraffic: scaleTraffic
    )
    let observation = SharedCoreTrafficRouteObservation(
      format: format,
      source: result.source,
      status: routeStatus(for: result.status)
    )
    report(observation)
    return result.text
  }

  static func resolveLazy(
    bytes: UInt64,
    nativeFallback: () -> String,
    format: SharedCoreTrafficFormat,
    abiVersion: () -> UInt32 = { MihomoMeterSharedCoreAdapter.abiVersion },
    scaleTraffic: (UInt64) throws -> SharedTrafficScale =
      MihomoMeterSharedCoreAdapter.scaleTraffic(bytes:)
  ) -> String {
    let result = SharedCoreTrafficRouter.routeLazy(
      bytes: bytes,
      nativeFallback: nativeFallback,
      format: format,
      abiVersion: abiVersion,
      scaleTraffic: scaleTraffic
    )
    report(
      SharedCoreTrafficRouteObservation(
        format: format,
        source: result.source,
        status: result.status
      )
    )
    return result.text
  }

  private static func report(_ observation: SharedCoreTrafficRouteObservation) {
    let reporter = stateLock.withLock { state -> Reporter? in
      guard let reporter = state.reporter,
        state.observationGate.shouldReport(observation)
      else {
        return nil
      }
      return reporter
    }
    do {
      try reporter?(observation)
    } catch {
      // 路由诊断不得改变共享主路径或原生回退结果。
    }
  }

  private static func routeStatus(
    for shadowStatus: SharedCoreTrafficShadowStatus
  ) -> SharedCoreTrafficRouteStatus {
    switch shadowStatus {
    case .matched:
      .matched
    case .abiMismatch:
      .abiMismatch
    case .nativeCallFailed:
      .nativeCallFailed
    case .unsupportedTrafficUnit, .unexpectedResult:
      .unexpectedResult
    case .mismatch:
      .mismatch
    case .unknownFailure:
      .unknownFailure
    }
  }
}
